javaslang/src/main/java/javaslang/collection/AbstractsQueue.java

```diff
@@ -0,0 +1,100 @@
+package javaslang.collection;
+
+import javaslang.*;
+import javaslang.control.Option;
+
+import java.util.*;
+
+abstract class AbstractsQueue<T, Q extends AbstractsQueue<T, Q>> implements Kind1<Q, T>, Traversable<T> {
```

Comment:
How should I make `PQ` final, as currently I need an `Empty` instance so that it throws a different exception for  `head` and `tail` (the tests asserted it)?